### PR TITLE
Some refactoring about tensor properties, make them handled more gene…

### PR DESCRIPTION
**Pull request by danpovey**
…rally, and make filler symbol configurable

### CODE REVIEW

File: k2/python/k2/fsa.py [hunk 1]
```diff
@@ -259,6 +259,28 @@ def __str__(self) -> str:
...
+            fsa.foo_filler = -1
+        which will mean the "filler" for attribute fsa.foo is -1; and
+        this will get propagated when you do FSA operations, like any
+        other non-tensor attribute.  The filler is the value that means
+        "nothing is here" (like epsilon).
+        Caution: you should use a value that is castable to float and back
+        to integer without loss of precision, because currently the
+        `default_value` parameter of `index_select` in ./ops.py is a float.
+        '''
+        try:
```
Comment (csukuangfj) [inline k2/python/k2/fsa.py:276]: How about
```python
assert attribute_name != 'aux_labels', 'you cannot set the filler for aux_labels'

ans = getattr(self, attribute_name + '_filler', 0)

return ans
```

Comment (danpovey) [reply k2/python/k2/fsa.py:276]: This doesn't work because we still *get* the filler for 'aux_labels', we just expect it to not be set.  Will change it, though, to also allow it to be set to 0.

Comment (csukuangfj) [reply k2/python/k2/fsa.py:276]: In that case, 
```python
ans = getattr(self, attribute_name + '_filler', 0) 
assert attribute_name != 'aux_labels' or ans == 0, 'you cannot set the filler for aux_labels' 
return ans
```
My intention is that

- (1) You don't need to use `try ... except` here. `getattr` accepts a default value

- (2) I usually see people using `assert some_expr, "some message"` in Python.
People in C/C++ use `assert(some_expr && "some message");`

File: k2/python/k2/fsa.py [hunk 2]
```diff
@@ -791,6 +821,55 @@ def requires_grad_(self, requires_grad: bool) -> 'Fsa':
...
+           It is OK if src_name and/or dest_name equals 'labels' or 'scores',
+           but these special attributes won't be deleted.
+        '''
+        assert src_name != dest_name
+        try:
+            value = getattr(self, src_name)
+            if src_name == 'labels':
+                value = value.clone()
+            setattr(self, dest_name, value)
+            try:
```
Comment (csukuangfj) [inline k2/python/k2/fsa.py:849]: If `src_name` is `scores`, you're deleting it, which does not match the `NOTE` given
in the comment.

File: k2/python/k2/fsa.py [hunk 3]
```diff
@@ -791,6 +821,55 @@ def requires_grad_(self, requires_grad: bool) -> 'Fsa':
...
+                deleted.
+        Returns: `self`
+
+        Note:
+           It is OK if src_name and/or dest_name equals 'labels' or 'scores',
+           but these special attributes won't be deleted.
+        '''
+        assert src_name != dest_name
+        try:
+            value = getattr(self, src_name)
```
Comment (csukuangfj) [inline k2/python/k2/fsa.py:845]: The implementation does not enforce that `src_name` is from the tensor attribute,
though the comment requires that it MUST be a tensor attribute.

Users may use it incorrectly.

File: k2/python/k2/utils.py [hunk 4]
```diff
@@ -440,7 +447,8 @@ def fsa_from_unary_function_tensor(src: Fsa, dest_arcs: _k2.RaggedArc,
 
 
 def fsa_from_unary_function_ragged(src: Fsa, dest_arcs: _k2.RaggedArc,
-                                   arc_map: _k2.RaggedInt) -> Fsa:
+                                   arc_map: _k2.RaggedInt,
+                                   remove_filler: bool = True) -> Fsa:
```
Comment (csukuangfj) [inline k2/python/k2/utils.py:451]: This newly added argument is not documented at all.

File: k2/python/k2/fsa.py (hunk 3 above)
Comment (danpovey) [reply k2/python/k2/fsa.py:845]: OK, I will check. (We also allow 'labels' as a special case, which I'll note).
